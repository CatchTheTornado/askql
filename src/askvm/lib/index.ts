export * from './extend';
export * from './resource';
export * from './run';
export * from './type';
export * from './typed';
export * from './preventReservedKeywords';
