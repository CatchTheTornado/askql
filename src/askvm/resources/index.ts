export * from './core';
export * from './list';
export * from './object';
export * from './math';
export * from './node';
export * from './string';
