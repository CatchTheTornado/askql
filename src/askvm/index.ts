import type { AskCodeOrValue } from '../askcode';
import { extendOptions, run as libRun, untyped } from './lib';
import type { JSONable, Options, TypedValue } from './lib';
import * as types from './lib/type';
import * as resources from './resources';
export { extendOptions, resource, typed, untyped } from './lib';
export type { Options, Resource, Resources, Values } from './lib';
export { resources, types };

export async function run(
  inputOptions: Options,
  code: AskCodeOrValue,
  args?: any[]
): Promise<TypedValue<JSONable>> {
  const options = extendOptions(inputOptions, {
    stats: {
      steps: 0,
    },
    limits: {
      steps: 10 ** 4,
    },
  });
  const result = await libRun(options, code, args);
  // console.log('stats after run', options.stats);
  return result;
}

export async function runUntyped(
  options: Options,
  code: AskCodeOrValue,
  args?: any[]
): Promise<JSONable> {
  return untyped(await run(options, code, args));
}
