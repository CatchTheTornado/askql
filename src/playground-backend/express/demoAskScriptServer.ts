import bodyParser from 'body-parser';
import cors from 'cors';
import express from 'express';

import { askCodeToSource } from '../../askcode';
import { resources as builtInResources, runUntyped } from '../../askvm';
import { parse as parseAskScript, AskScriptCode } from '../../askscript';

import chalk = require('chalk');
import { customAlphabet } from 'nanoid';
import { customResources } from '../lib/resources';
import { customValues } from '../lib/values';
import { compileAskCode, logError } from '../lib/utils';

const packageInfo = require('../../../package.json');

const nanoid = customAlphabet('1234567890abcdef', 8);

const baseEnvironment = {
  resources: { ...builtInResources, ...customResources },
  values: customValues,
};

const app = express();
app.use(cors());
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));

app.get('/', async (req, res) => {
  res.type('html');
  res.send(
    '<html><head><title>AskQL</title></head><body><p><b>Ask</b> me anything!</p><p><a href="http://askql.org/">AskQL website</a></p></body></html>'
  );
});

app.get('/version', async (req, res) => {
  res.json({ version: packageInfo.version });
});

app.post('/askscript', async (req, res) => {
  function reportError(e: Error, message: string) {
    res.status(400).json({
      message,
      error: e.toString(),
    });
  }

  const code: AskScriptCode = req.body.code;

  if (typeof code !== 'string') {
    res.status(400).json({
      message: 'Please send Askscript code in "code" property',
    });
    return;
  }

  let askCode;
  let askCodeSource;

  const id = nanoid();

  try {
    console.log(id + ' -- ' + new Date().toString());
    console.log(id + ' -- ' + chalk.grey(`➡️ ${code}`));

    askCode = parseAskScript(code);

    askCodeSource = askCodeToSource(askCode);
  } catch (e) {
    logError(id, code, e);

    reportError(e, 'Could not compile your AskScript code');
    return;
  }

  try {
    const result = await compileAskCode(baseEnvironment, askCode);

    console.log(id + ' -- ' + chalk.grey(`⬅️ ${JSON.stringify(result)}`));
    console.log('\n\n');
    res.json({
      askCodeSource,
      message: 'Code run successfully',
      result,
    });
  } catch (e) {
    logError(id, code, e);

    reportError(e, 'Could not run your code');
  }
});

const port = process.env.PORT || 4000;
app.listen(port, () => {
  console.log(chalk.grey(`AskQL listening at http://localhost:${port}`));
});
