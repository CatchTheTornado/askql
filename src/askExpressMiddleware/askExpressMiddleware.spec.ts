import {
  askExpressMiddleware,
  AskExpressMiddlewareConfig,
} from './askExpressMiddleware';
import { NextFunction } from 'express';
import { resources } from '../askvm';

describe(`askExpressMiddleware`, () => {
  const values = {
    hello: 'Hello world, this is AskQL ðŸ¦„',
  };

  const config: AskExpressMiddlewareConfig = {
    callNext: true,
    passError: false,
  };

  let middlware, mockRequest, mockResponse, mockNext;

  beforeEach(() => {
    middlware = askExpressMiddleware({ values, resources });
    mockNext = jest.fn() as jest.Mock<NextFunction>;
    mockRequest = ({
      body: {},
    } as unknown) as jest.Mock<Request>;
    mockResponse = ({
      json: jest.fn(),
    } as unknown) as jest.Mock<Response>;
  });

  it('should respond with the queried resource', async (done) => {
    // const askScript = `ask(query(node('hello',f(call(get('toUpperCase'),call(get('get'),'hello'))))))`;
    const askScript = `ask(f(call(get('toUpperCase'), call(get('get'),'hello'))))`;
    const expectedValue = values.hello.toUpperCase();

    mockRequest.body.code = askScript;

    await middlware(mockRequest, mockResponse, mockNext);
    expect(mockResponse.json).toHaveBeenCalledWith(expectedValue);
    done();
  });

  it('should run next without arguments', () => {
    middlware(mockRequest, mockResponse, mockNext);

    expect(mockNext).toHaveBeenCalledWith();
  });

  it('should not run next if passed callNext false', () => {
    const errorProneMiddleware = askExpressMiddleware({}, { callNext: false });
    errorProneMiddleware(mockRequest, mockResponse, mockNext);
    expect(mockNext).not.toHaveBeenCalled();
  });

  it('should run next with error when passed the error argument', () => {
    const errorProneMiddleware = askExpressMiddleware(
      {},
      { callNext: true, passError: true }
    );
    mockRequest.body = undefined; // should cause an error
    const expectedError = new TypeError(
      `Cannot read property 'code' of undefined`
    );
    errorProneMiddleware(mockRequest, mockResponse, mockNext);
    expect(mockNext).toHaveBeenCalledWith(expectedError);
  });
});
