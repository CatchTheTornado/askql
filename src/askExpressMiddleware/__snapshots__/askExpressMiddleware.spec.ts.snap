// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`askExpressMiddleware should call next with error when bad code is sent 1`] = `
[SyntaxError: Expected all operators in an expression to be the same (e.g. a+b+c), but 5 + 5 * 10, which uses operator '*', found instead. Use brackets to separate different operators, e.g. a+b+(c*d)
Location: {
  "start": {
    "offset": 12,
    "line": 2,
    "column": 7
  },
  "end": {
    "offset": 22,
    "line": 2,
    "column": 17
  }
}]
`;
